crates/span/src/ast_id.rs

```diff
@@ -152,126 +548,345 @@ impl AstIdMap {
+        res.arena.alloc((SyntaxNodePtr::new(node), ROOT_ERASED_FILE_AST_ID));
 
-        // make sure to allocate the root node
-        if !should_alloc_id(node.kind()) {
-            res.alloc(node);
-        }
         // By walking the tree in breadth-first order we make sure that parents
         // get lower ids then children. That is, adding a new child does not
         // change parent's id. This means that, say, adding a new function to a
         // trait does not change ids of top-level items, which helps caching.
-        bdfs(node, |it| {
-            if should_alloc_id(it.kind()) {
-                res.alloc(&it);
-                TreeOrder::BreadthFirst
-            } else {
-                TreeOrder::DepthFirst
+
+        // This contains the stack of the `BlockExpr`s we are under. We do this
+        // so we only allocate `BlockExpr`s if they contain items.
+        // The general idea is: when we enter a block we push `(block, false)` here.
+        // Items inside the block are attributed to the block's container, not the block.
+        // For the first item we find inside a block, we make this `(block, true)`
+        // and create an ast id for the block. When exiting the block we pop it,
+        // whether or not we created an ast id for it.
+        // It may seem that with this setup we will generate an ID for blocks that
+        // have no items directly but have items inside other items inside them.
+        // This is true, but it doesn't matter, because such blocks can't exist.
+        // After all, the block will then contain the *outer* item, so we allocate
+        // an ID for it anyway.
+        let mut blocks = Vec::new();
+        let mut curr_layer = vec![(node.clone(), None)];
+        let mut next_layer = vec![];
+        while !curr_layer.is_empty() {
+            curr_layer.drain(..).for_each(|(node, parent_idx)| {
+                let mut preorder = node.preorder();
+                while let Some(event) = preorder.next() {
+                    match event {
+                        syntax::WalkEvent::Enter(node) => {
+                            if ast::BlockExpr::can_cast(node.kind()) {
+                                blocks.push((node, false));
+                            } else if ErasedFileAstId::should_alloc(&node) {
+                                // Allocate blocks on-demand, only if they have items.
+                                // We don't associate items with blocks, only with items, since block IDs can be quite unstable.
+                                // FIXME: Is this the correct thing to do? Macro calls might actually be more incremental if
+                                // associated with blocks (not sure). Either way it's not a big deal.
+                                if let Some((last_block_node, already_allocated)) =
+                                    blocks.last_mut()
+                                {
+                                    if !*already_allocated {
```

Comment:
```suggestion
                                if let Some((last_block_node, already_allocated @ false)) =
                                    blocks.last_mut()
                                {
```
This might look worse depending on rustfmt so feel free to ignore (also binding modes might not work in favor here so this suggestion could be wrong, not entirely sure right now)